Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,646 @@
+      fun c => ⟨ofSubtype
+          (p.1.subtypePerm (Classical.choose (hp c.val c.prop))),
+            Classical.choose_spec (hp c.val c.prop)⟩
+    use (u, v)
+    ext x
+    rw [θHom_apply]
+    split_ifs with hx
+    · rw [cycleOf_mem_cycleFactorsFinset_iff, mem_support] at hx
+      rw [ofSubtype_apply_of_mem, subtypePerm_apply]
+      rwa [mem_support, cycleOf_apply_self, ne_eq]
+    · rw [cycleOf_mem_cycleFactorsFinset_iff, not_mem_support] at hx
+      rwa [ofSubtype_apply_of_mem, subtypePerm_apply]
+
+theorem θHom_range_card (g : Equiv.Perm α) :
+    Fintype.card (θHom g).range = (Fintype.card α - g.cycleType.sum)! * g.cycleType.prod := by
+  erw [Set.card_range_of_injective (θHom_injective g)]
+  rw [Fintype.card_prod, Fintype.card_perm, Fintype.card_pi, card_fixedPoints]
+  apply congr_arg
+  rw [Finset.univ_eq_attach, g.cycleFactorsFinset.prod_attach (fun i ↦ Fintype.card (zpowers i)),
+    cycleType, Finset.prod_map_val]
+  refine Finset.prod_congr rfl (fun x hx ↦ ?_)
+  rw [Fintype.card_zpowers, (mem_cycleFactorsFinset_iff.mp hx).1.orderOf, Function.comp_apply]
+
+end Kernel
+
+end OnCycleFactors
+
+open Nat
+
+variable (g : Perm α)
+
+-- Should one parenthesize the product ?
+/-- Cardinality of the centralizer in `Equiv.Perm α` of a permutation given `cycleType` -/
+theorem nat_card_centralizer :
+    Nat.card (centralizer {g}) =
+      (Fintype.card α - g.cycleType.sum)! * g.cycleType.prod *
+        (∏ n in g.cycleType.toFinset, (g.cycleType.count n)!) := by
+  rw [← (toPermHom g).ker.card_mul_index, index_ker, nat_card_range_toPermHom,
+    ← θHom_range_card, ← Nat.card_eq_fintype_card, θHom_range_eq, card_subtype]
+
+theorem card_isConj_mul_eq :
+    Nat.card {h : Perm α | IsConj g h} *
+      ((Fintype.card α - g.cycleType.sum)! *
+      g.cycleType.prod *
+      (∏ n in g.cycleType.toFinset, (g.cycleType.count n)!)) =
+    (Fintype.card α)! := by
+  classical
+  rw [Nat.card_eq_fintype_card, ← nat_card_centralizer g]
+  rw [Subgroup.nat_card_centralizer_nat_card_stabilizer, Nat.card_eq_fintype_card]
+  convert MulAction.card_orbit_mul_card_stabilizer_eq_card_group (ConjAct (Perm α)) g
+  · ext h
+    simp only [Set.mem_setOf_eq, ConjAct.mem_orbit_conjAct, isConj_comm]
+  · rw [ConjAct.card, Fintype.card_perm]
+
+/-- Cardinality of a conjugacy class in `Equiv.Perm α` of a given `cycleType` -/
+theorem card_isConj_eq :
+    Nat.card {h : Perm α | IsConj g h} =
+      (Fintype.card α)! /
+        ((Fintype.card α - g.cycleType.sum)! *
+          g.cycleType.prod *
+          (∏ n in g.cycleType.toFinset, (g.cycleType.count n)!)) := by
+  rw [← card_isConj_mul_eq g, Nat.div_eq_of_eq_mul_left _]
+  · rfl
+  -- This is the cardinal of the centralizer
+  · rw [← nat_card_centralizer g]
+    apply Nat.card_pos
+
+variable (α)
+
+theorem card_of_cycleType_eq_zero_iff {m : Multiset ℕ} :
+    (Finset.univ.filter fun g : Equiv.Perm α => g.cycleType = m).card = 0
```

Comment:
Here and below, consider using the new Finset univ elaborator:
```
    ({g | g.cycleType = m} : Finset (Perm α)).card 
```

Reply:
Done.